geomstats/geometry/complex_poincare_disk.py

```diff
@@ -0,0 +1,360 @@
+.. [JV2016] B. Jeuris and R. Vandebril. The Kahler mean of Block-Toeplitz
+    matrices with Toeplitz structured blocks, 2016.
+    https://epubs.siam.org/doi/pdf/10.1137/15M102112X
+.. [Yang2013] Marc Arnaudon, Frédéric Barbaresco and Le Yang.
+    Riemannian medians and means with applications to radar signal processing,
+    IEEE Journal of Selected Topics in Signal Processing, vol. 7, no. 4, pp. 595-604,
+    Aug. 2013, doi: 10.1109/JSTSP.2013.2261798.
+    https://ieeexplore.ieee.org/document/6514112
+"""
+
+import geomstats.backend as gs
+from geomstats.geometry.base import ComplexOpenSet
+from geomstats.geometry.complex_riemannian_metric import ComplexRiemannianMetric
+from geomstats.geometry.hermitian import Hermitian
+
+
+class ComplexPoincareDisk(ComplexOpenSet):
+    """Class for the complex Poincaré disk.
+
+    The Poincaré disk is a representation of the Hyperbolic
+    space of dimension 2. Its complex dimension is 1.
+
+    Parameters
+    ----------
+    scale : float
+        Scale of the complex Poincaré metric.
+        Optional, default: 1.
+    """
+
+    def __init__(self, scale=1.0, **kwargs):
+        super().__init__(
+            dim=1,
+            embedding_space=Hermitian(dim=1),
+            metric=ComplexPoincareDiskMetric(scale=scale),
+            **kwargs
+        )
+        self.scale = scale
+
+    @staticmethod
+    def belongs(point, atol=gs.atol):
+        """Check if a point belongs to the complex unit disk.
+
+        Check if a point belongs to the Poincaré complex disk,
+        i.e. Check if its norm is lower than one.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., 1]
+            Point to be checked.
+        atol : float
+            Tolerance.
+            Optional, default: backend atol.
+
+        Returns
+        -------
+        belongs : array-like, shape=[...,]
+            Boolean denoting if point belongs to the
+            complex Poincaré disk.
+        """
+        is_scalar = gs.ndim(point) == 1
+        is_scalar = is_scalar or (gs.ndim(point) == 2 and point.shape[1] == 1)
+        has_modulus_lower_than_one = gs.array(
+            [[gs.all(gs.abs(point[i_point]) < 1)] for i_point in range(point.shape[0])]
+        )
+        belongs = gs.logical_and(is_scalar, has_modulus_lower_than_one)
+        belongs = gs.reshape(belongs, (-1,))
+        return belongs
+
+    @staticmethod
+    def projection(point):
+        """Project a point on the complex unit disk.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., 1]
+            Point to project.
+        atol : float
+            Tolerance.
+            Optional, default: backend atol.
+
+        Returns
+        -------
+        projected : array-like, shape=[..., 1]
+            Projected point.
+        """
+        return gs.where(
+            gs.abs(point) >= 1 - gs.atol,
+            gs.cast((1 - gs.atol) / gs.abs(point), dtype=point.dtype) * point,
```

Comment:
`cast` is unnecessary here.

Reply:
If I do not cast, TensorFlow gives the following errors:

```
========================================================================================================== FAILURES ===========================================================================================================
__________________________________________________________________________________ TestComplexPoincareDisk.test_projection[point0-expected0] __________________________________________________________________________________

self = <tests.tests_geomstats.test_complex_poincare_disk.TestComplexPoincareDisk object at 0x7fa18ef6c7f0>, point = [(1+0j), 0j], expected = [(0.999999+0j), 0j]

    def test_projection(self, point, expected):
        self.assertAllClose(
>           self.Space().projection(gs.array(point)),
            gs.array(expected),
        )

tests/tests_geomstats/test_complex_poincare_disk.py:24: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
geomstats/geometry/complex_poincare_disk.py:96: in projection
    (1 - gs.atol) / gs.abs(point) * point,
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1407: in binary_op_wrapper
    return func(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1767: in _mul_dispatch
    return multiply(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py:1176: in op_dispatch_handler
    return dispatch_target(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:529: in multiply
    return gen_math_ops.mul(x, y, name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/gen_math_ops.py:6581: in mul
    _ops.raise_from_not_ok_status(e, name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

e = _NotOkStatusException(), name = None

    def raise_from_not_ok_status(e, name):
      e.message += (" name: " + name if name is not None else "")
>     raise core._status_to_exception(e) from None  # pylint: disable=protected-access
E     tensorflow.python.framework.errors_impl.InvalidArgumentError: cannot compute Mul as input #1(zero-based) was expected to be a double tensor but is a complex128 tensor [Op:Mul]

venv/lib/python3.8/site-packages/tensorflow/python/framework/ops.py:7215: InvalidArgumentError
__________________________________________________________________________________ TestComplexPoincareDisk.test_projection[point1-expected1] __________________________________________________________________________________

self = <tests.tests_geomstats.test_complex_poincare_disk.TestComplexPoincareDisk object at 0x7fa18ef6c880>, point = [(-0-1j), (-2+2j)], expected = [(-0-0.999999j), (-0.7071060740797662+0.7071060740797662j)]

    def test_projection(self, point, expected):
        self.assertAllClose(
>           self.Space().projection(gs.array(point)),
            gs.array(expected),
        )

tests/tests_geomstats/test_complex_poincare_disk.py:24: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
geomstats/geometry/complex_poincare_disk.py:96: in projection
    (1 - gs.atol) / gs.abs(point) * point,
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1407: in binary_op_wrapper
    return func(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1767: in _mul_dispatch
    return multiply(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py:1176: in op_dispatch_handler
    return dispatch_target(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:529: in multiply
    return gen_math_ops.mul(x, y, name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/gen_math_ops.py:6581: in mul
    _ops.raise_from_not_ok_status(e, name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

e = _NotOkStatusException(), name = None

    def raise_from_not_ok_status(e, name):
      e.message += (" name: " + name if name is not None else "")
>     raise core._status_to_exception(e) from None  # pylint: disable=protected-access
E     tensorflow.python.framework.errors_impl.InvalidArgumentError: cannot compute Mul as input #1(zero-based) was expected to be a double tensor but is a complex128 tensor [Op:Mul]

venv/lib/python3.8/site-packages/tensorflow/python/framework/ops.py:7215: InvalidArgumentError
====================================================================================================== warnings summary =======================================================================================================
venv/lib/python3.8/site-packages/flatbuffers/compat.py:19
  /home/ycabanes/work/geomstats/geomstats/venv/lib/python3.8/site-packages/flatbuffers/compat.py:19: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

venv/lib/python3.8/site-packages/tensorflow_probability/python/__init__.py:57
  /home/ycabanes/work/geomstats/geomstats/venv/lib/python3.8/site-packages/tensorflow_probability/python/__init__.py:57: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    if (distutils.version.LooseVersion(tf.__version__) <

venv/lib/python3.8/site-packages/tensorflow_probability/python/__init__.py:58
  /home/ycabanes/work/geomstats/geomstats/venv/lib/python3.8/site-packages/tensorflow_probability/python/__init__.py:58: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    distutils.version.LooseVersion(required_tensorflow_version)):

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=================================================================================================== short test summary info ===================================================================================================
FAILED tests/tests_geomstats/test_complex_poincare_disk.py::TestComplexPoincareDisk::test_projection[point0-expected0] - tensorflow.python.framework.errors_impl.InvalidArgumentError: cannot compute Mul as input #1(zero-b...
FAILED tests/tests_geomstats/test_complex_poincare_disk.py::TestComplexPoincareDisk::test_projection[point1-expected1] - tensorflow.python.framework.errors_impl.InvalidArgumentError: cannot compute Mul as input #1(zero-b...
==================================================================================== 2 failed, 41 passed, 19 skipped, 3 warnings in 1.03s =====================================================================================
```